File: Searching Algorithms/binary_search.c
/*
Binary Search

input
5
1 2 3 4 5
3

output
2

For binary search algorithm to work, the array should be in a sorted order. This search algorithm returns
the index of the element that is found. If the element is not found it returns -1.

*/
#include<stdio.h>

int binarySearch(int a[],int n,int k)
{
  int first = 0;
  int last = n-1;
  while(first <= last)
  {
    int mid = (first + last)/2;
    if(a[mid]>k)
    last = mid -1;
    else if(a[mid]<k)
    first = mid + 1;
    else if(a[mid]==k)
      return mid;
  }
  return -1;
}
int main(int argc, char const *argv[])
{
    int n,k,pos;
    scanf("%d",&n);
    int a[n];;
    for(int i=0;i<n;i++)
    {
        scanf("%d",&a[i]);
    }
    scanf("%d",&k);
    pos = binarySearch(a,n,k);
    if(pos == -1)
      printf("Element is not found!");
    else
      printf("Element %d is found at %d",k,pos);
    return 0;
}

File: Searching Algorithms/binary_search.py
# BINARY SEARCH 

# It is a searching algorithm which is more efficient than linear search. It can find an element if present in O(log(n)) time complexity.
# For this algorithm to work, the array should be in sorted order
# it returns the position of the element in the array


def binary_search (arr, item):
    """The binary_Search function takes input the sorted list of elements and the item to be searched
        and returns the position of the item, if found in the list, else it returns -1 """
    beg, end = 0, len(arr)-1
    mid = (beg+end)//2

    while beg<=end:
        mid = (beg+end)//2
        if arr[mid] == item :
            return mid
        elif arr[mid] > item:
            end = mid-1
        elif arr[mid] < item:
            beg = mid+1
    return -1                   # return -1 if the element is not present in the array



## DRIVER CODE

test_arr = [1,2,3,4,5]      # this should be a sorted array
test_item = 2               # this is the item to be searched

ans = binary_search(test_arr, test_item)
if ans != -1 :
    print(f"Element found at index position {ans}")
else:
    print("Element is not present in array")